src/celeritas/ext/detail/BVHNavigator.hh

```diff
@@ -99,21 +102,29 @@ class BVHNavigator
     // out_state.SetBoundaryState(true) and hitcandidate is set to the hit
     // daughter volume, or kept unchanged if the current volume is left.
     CELER_FUNCTION static double
-    ComputeStepAndHit(vecgeom::Vector3D<vecgeom::Precision> const& localpoint,
-                      vecgeom::Vector3D<vecgeom::Precision> const& localdir,
-                      vecgeom::Precision                           step_limit,
-                      vecgeom::NavigationState const&              in_state,
-                      vecgeom::NavigationState&                    out_state,
-                      VPlacedVolumePtr_t& hitcandidate)
+    ComputeStepAndHit(Vector3D const&                 localpoint,
+                      Vector3D const&                 localdir,
+                      Precision                       step_limit,
+                      vecgeom::NavigationState const& in_state,
+                      vecgeom::NavigationState&       out_state,
+                      VPlacedVolumePtr_t&             hitcandidate)
     {
-        vecgeom::Precision step = step_limit;
+        if (step_limit <= 0.)
+        {
+            // We don't need to ask any solid, step not limited by geometry.
+            in_state.CopyTo(&out_state);
+            out_state.SetBoundaryState(false);
+            return 0.;
```

Comment:
All of these literal `0.` are doubles, which can promote comparisons and other operations on single-precision values to double precision. This is why we've been careful in Celeritas to use `0` rather than `0.` for floating point operations: int gets promoted to whatever real type you're using, but operations with `0.` will always be promoted to double.

Since the value `double{0}` is exactly representable as float, the compiler seems to be able to optimize this, but I think it's bad practice to rely on the compiler optimization to downcast, when you're "telling" it (by giving a double-precision argument) to upcast.